File: kickstartF2021/festival2.py
import math

DEV = True

sample_text = """2
10 4 2
800 2 8
1500 6 9
200 4 7
400 3 5
5 3 3
400 1 3
500 5 5
300 2 3
"""


if DEV:
	from collections import deque
	from unittest.mock import MagicMock
	sample_lines = deque(sample_text.split("\n"))

	input = MagicMock(side_effect=lambda : sample_lines.popleft())



def parse_input():
	n_test_cases = int(input())
	for t in range(n_test_cases):
		D, N, K = list(map(int, input().split()))
		rides = []
		for _ in range(N):
			h, s, e = list(map(int, input().split()))
			rides.append((h, s, e))
		yield (D, K, rides)

class SegmentTree(object):
	def __init__(self, N,
				 query_fn=lambda x, y: x + y,
				 update_fn=lambda x, y: y,
				 default_val=0):

		self.original_N = N
		intended_length = 2**(math.ceil(math.log2(N))+1)
		self.N = intended_length
		self.H = (self.N-1).bit_length()
		self.query_fn = query_fn
		self.update_fn = update_fn
		self.default_val = default_val
		self.tree = [default_val] * (2 * self.N)
		self.lazy = [None] * self.N

	def __apply(self, x, val):
		self.tree[x] = self.update_fn(self.tree[x], val)
		if x < self.N:
			self.lazy[x] = self.update_fn(self.lazy[x], val)

	def bulk_update(self, I, hs):
		for i, h in enumerate(hs):
			self.update(I+i, I+i, h)

	def update(self, L, R, h):
		def pull(x):
			while x > 1:
				x //= 2
				self.tree[x] = self.query_fn(self.tree[x*2], self.tree[x*2 + 1])
				if self.lazy[x] is not None:
					self.tree[x] = self.update_fn(self.tree[x], self.lazy[x])
		L += self.N
		R += self.N
		L0, R0 = L, R
		while L <= R:
			if L & 1:
				self.__apply(L, h)
				L += 1
			if R & 1 == 0:
				self.__apply(R, h)
				R -= 1
			L //= 2
			R //= 2
		pull(L0)
		pull(R0)

	def query(self, L, R):
		def push(x):
			n = 2**self.H
			while n != 1:
				y = x // n
				if self.lazy[y] is not None:
					self.__apply(y*2, self.lazy[y])
					self.__apply(y*2 + 1, self.lazy[y])
					self.lazy[y] = None
				n //= 2

		result = self.default_val
		if L > R:
			return result

		L += self.N
		R += self.N
		push(L)
		push(R)
		while L <= R:
			if L & 1:
				result = self.query_fn(result, self.tree[L])
				L += 1
			if R & 1 == 0:
				result = self.query_fn(result, self.tree[R])
				R -= 1
			L //= 2
			R //= 2
		return result

	def idx_at_or_below(self, k):
		v = 1
		tl = 0
		tr = self.N-1

		while tl != tr:
			if k > self.tree[v]:
				return self.original_N 
			tm = (tl + tr) // 2
			if self.tree[v*2] > k:
				v *= 2
				tr = tm
			else:
				k -= self.tree[v*2]
				v = v * 2 + 1
				tl = tm + 1

		return tl - 1
	
	def data(self):
		showList = []
		for i in xrange(self.N):
			showList.append(self.query(i, i))
		return showList

def test_segment_tree():
	arrays = [
		[1,2,3,4],
		[1,2,3,4,5],
		[1,2,3,4,5,6],
	]

	for a in arrays:
		tree = SegmentTree(len(a))
		tree.bulk_update(0, a)
		idx = tree.idx_at_or_below(-1000)
		assert idx == -1, idx

		idx = tree.idx_at_or_below(1)
		assert idx == 0, idx

		idx = tree.idx_at_or_below(2)
		assert idx == 0, idx

		idx = tree.idx_at_or_below(3)
		assert idx == 1, idx

		idx = tree.idx_at_or_below(5)
		assert idx == 1, idx

		idx = tree.idx_at_or_below(6)
		assert idx == 2, idx

		idx = tree.idx_at_or_below(1000)
		assert idx == len(a), idx

		s = tree.query(2, 2)
		assert s == 3, s

		s = tree.query(0, 1)
		assert s == 3, s

		s = tree.query(1, 3)
		assert s == 9, s


if DEV:
	test_segment_tree()

def solution(D, K, rides):
	line_sweep = []
	hs = []
	for i, (h, s, e) in enumerate(rides):
		line_sweep.append((s, 1, h, i))
		line_sweep.append((e+1, -1, h, i))
		hs.append((h, i))

	hs.sort(reverse=True)
	line_sweep.sort()

	idx_to_rank = {i: rank for rank, (_, i) in enumerate(hs)}

	enabled = SegmentTree(len(rides))
	scores = SegmentTree(len(rides))

	result = 0
	for _, type_, h, i in line_sweep:
		if type_ == 1:
			enabled.update(idx_to_rank[i], idx_to_rank[i], 1)
			scores.update(idx_to_rank[i], idx_to_rank[i], h)
			idx = enabled.idx_at_or_below(K)
			if idx == -1:
				idx = len(rides) 
			
			result = max(result, scores.query(0, idx))
		else:
			enabled.update(idx_to_rank[i], idx_to_rank[i], 0)
			scores.update(idx_to_rank[i], idx_to_rank[i], 0)

	return result



def main():
	for i, test_case in enumerate(parse_input()):
		print("Case #{}: {}".format(i+1, solution(*test_case)))

if __name__ == '__main__':
	main()

File: kickstartF2021/festival2.go
package main

import (
    "math"
    "bufio"
    "fmt"
    "os"
	"sort"
)


type SegmentTree struct {
    nodes []int
    size int
    oSize int
}


func calcTreeSize(originalSize int) int {
    return 1<<uint(math.Ceil(math.Log2(float64(originalSize)))+1)
}

func NewSegmentTree(s int) *SegmentTree{
    size := calcTreeSize(s)
    nodes := make([]int, size*2)

    t := &SegmentTree{nodes, size, s}
    return t
}

func NewSegmentTreeFromSlice(from []int) *SegmentTree{
    size := calcTreeSize(len(from))
    nodes := make([]int, size*4)

    t := &SegmentTree{nodes, size, len(from)}
    for i := 0; i < len(from); i += 1 {
        nodes[i+size] = from[i]
    }
    for i := size - 1; i > 0; i -= 1 {
        t.nodes[i] = t.nodes[i<<1] + t.nodes[i<<1|1];
    }
    return t
}

func (t *SegmentTree) query(left, right int) int {
    res := 0
    left += t.size
    right += t.size;
    for left < right  {
        if left&1 == 1 {
            res += t.nodes[left]
            left += 1
        }
        if right&1 == 1 {
            right -= 1
            res += t.nodes[right]
        }
        left >>= 1
        right >>= 1
    }
    return res
}

func (t *SegmentTree) update(idx, value int) {
    idx += t.size
    t.nodes[idx] = value
    for idx > 1 {
        t.nodes[idx>>1] = t.nodes[idx] + t.nodes[idx^1]
        idx -= 1
    }
}

func (t *SegmentTree) idxAtOrBelow(k int) int {
	v := 1
	tl := 0
	tr := t.size - 1

    var tm int
	for tl != tr {
		if k > t.nodes[v] {
            return t.oSize
        }
        tm = (tl + tr) / 2
        if t.nodes[v*2] > k {
            v *= 2
            tr = tm
        } else {
            k -= t.nodes[v*2]
            v = v * 2 + 1
            tl = tm + 1
        }
	}

    return tl - 1
}


	// def idx_at_or_below(self, k):
	// 	v = 1
	// 	tl = 0
	// 	tr = self.N-1
    //
	// 	while tl != tr:
	// 		if k > self.tree[v]:
	// 			return self.original_N 
	// 		tm = (tl + tr) // 2
	// 		if self.tree[v*2] > k:
	// 			v *= 2
	// 			tr = tm
	// 		else:
	// 			k -= self.tree[v*2]
	// 			v = v * 2 + 1
	// 			tl = tm + 1
    //
	// 	return tl - 1

func testSegmentTree() {
	for _, arr := range [][]int{[]int{1,2,3,4}, []int{1,2,3,4,5}, []int{1,2,3,4,5,6}} {
		tree := NewSegmentTreeFromSlice(arr)
		var s int
		s = tree.query(0, 3)
		fmt.Println(6, s)

		s = tree.query(1, 3)
		fmt.Println(5, s)

		tree.update(1, 0)
		s = tree.query(1, 3)
		fmt.Println(3, s)

		tree.update(1, 2)
		fmt.Println(tree.nodes)

		s = tree.query(1, 3)
		fmt.Println(5, s)

		s = tree.idxAtOrBelow(1)
		fmt.Println(0, s)

		s = tree.idxAtOrBelow(2)
		fmt.Println(0, s)

		s = tree.idxAtOrBelow(3)
		fmt.Println(1, s)

		s = tree.idxAtOrBelow(5)
		fmt.Println(1, s)

		s = tree.idxAtOrBelow(6)
		fmt.Println(2, s)

		s = tree.idxAtOrBelow(1000)
		fmt.Println(len(arr), s)

		s = tree.idxAtOrBelow(-1000)
		fmt.Println(-1, s)
	}
	fmt.Println("test done")
}

type Attraction struct {
    h int
    s int
    e int
}


func solution(D int, K int, attractions []Attraction) int {
    hs := make([][]int,  0)
    lineSweep := make([][]int, 0)
    for i, ride := range attractions {
        lineSweep = append(lineSweep, []int{ride.s, 1, ride.h, i})
        lineSweep = append(lineSweep, []int{ride.e+1, -1, ride.h, i})
        hs = append(hs, []int{ride.h, i})
    }
    sort.Slice(hs[:],func(i, j int) bool {
		for x := range hs[i] {
			if hs[i][x] == hs[j][x] {
				continue
			}
			return hs[i][x] > hs[j][x]
		}
		return false
	})
    sort.Slice(lineSweep[:],func(i, j int) bool {
		for x := range lineSweep[i] {
			if lineSweep[i][x] == lineSweep[j][x] {
				continue
			}
			return lineSweep[i][x] < lineSweep[j][x]
		}
		return false
	})

	idxToRank := make(map[int]int)
	for i, elt := range hs {
		idxToRank[elt[1]] = i
	}


	enabled := NewSegmentTree(len(attractions))
	scores := NewSegmentTree(len(attractions))

	result := 0
	for _, elt := range lineSweep {
		type_ := elt[1]
		h := elt[2]
		i := elt[3]
		if type_ == 1 {
			enabled.update(idxToRank[i], 1)
			scores.update(idxToRank[i], h)
			idx := enabled.idxAtOrBelow(K)
			if idx == -1 {
				idx = len(attractions)
			}
			idx += 1
			result = Max(result, scores.query(0, idx))
		} else {
			enabled.update(idxToRank[i], 0)
			scores.update(idxToRank[i], 0)
		}
	}


    return result
}

func Max(x, y int) int {
    if x < y {
        return y
    }
    return x
}

func main() {
	// testSegmentTree()
    in := bufio.NewReader(os.Stdin)

    var T int
    fmt.Fscan(in, &T)
    for i := 1; i <= T; i++ {
        var D, N, K int
        fmt.Fscan(in, &D, &N, &K)

        attractions := make([]Attraction, N)
        var h, s, e int
        for j := 0; j < N; j++ {
            fmt.Fscan(in, &h, &s, &e)
            attractions[j] = Attraction{h, s, e}
        }
        result := solution(D, K, attractions)
        fmt.Printf("Case #%d: %d\n", i, result)
    }
}
